File: aicode/minicli4/typescript/src/tools.ts
import childProcess from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { RuntimeConfig } from "./config";

export interface ToolDefinition {
  type: "function";
  function: {
    name: string;
    description: string;
    parameters: {
      type: "object";
      properties: Record<string, unknown>;
      required?: string[];
    };
  };
}

export interface ToolResult {
  ok: boolean;
  output: string;
  requiresApproval?: boolean;
}

function withinRoot(projectRoot: string, target: string): boolean {
  const root = path.resolve(projectRoot);
  const resolved = path.resolve(projectRoot, target);
  return resolved === root || resolved.startsWith(`${root}${path.sep}`);
}

function normalizeAllowed(config: RuntimeConfig, projectRoot: string): string[] {
  return config.allowed_paths.map((entry) => path.resolve(projectRoot, entry));
}

function allowedPath(config: RuntimeConfig, projectRoot: string, target: string): boolean {
  const resolved = path.resolve(projectRoot, target);
  const allowed = normalizeAllowed(config, projectRoot);
  return allowed.some((base) => resolved === base || resolved.startsWith(`${base}${path.sep}`));
}

function parseArgs(raw: unknown): Record<string, unknown> {
  if (typeof raw === "string") {
    try {
      const parsed = JSON.parse(raw) as Record<string, unknown>;
      return parsed;
    } catch {
      return {};
    }
  }
  if (raw && typeof raw === "object") {
    return raw as Record<string, unknown>;
  }
  return {};
}

export class ToolRegistry {
  constructor(private readonly projectRoot: string, private readonly config: RuntimeConfig) {}

  definitions(): ToolDefinition[] {
    return [
      {
        type: "function",
        function: {
          name: "read_file",
          description: "Read a UTF-8 text file.",
          parameters: {
            type: "object",
            properties: { path: { type: "string" } },
            required: ["path"]
          }
        }
      },
      {
        type: "function",
        function: {
          name: "write_file",
          description: "Write UTF-8 content to file.",
          parameters: {
            type: "object",
            properties: { path: { type: "string" }, content: { type: "string" } },
            required: ["path", "content"]
          }
        }
      },
      {
        type: "function",
        function: {
          name: "list_dir",
          description: "List directory entries.",
          parameters: {
            type: "object",
            properties: { path: { type: "string" } },
            required: ["path"]
          }
        }
      },
      {
        type: "function",
        function: {
          name: "grep_text",
          description: "Search project text with regex.",
          parameters: {
            type: "object",
            properties: { pattern: { type: "string" }, path: { type: "string" } },
            required: ["pattern"]
          }
        }
      },
      {
        type: "function",
        function: {
          name: "run_shell",
          description: "Run shell command under policy.",
          parameters: {
            type: "object",
            properties: { command: { type: "string" } },
            required: ["command"]
          }
        }
      },
      {
        type: "function",
        function: {
          name: "project_tree",
          description: "Render project tree.",
          parameters: {
            type: "object",
            properties: { path: { type: "string" }, depth: { type: "integer" } }
          }
        }
      },
      {
        type: "function",
        function: {
          name: "session_export",
          description: "Export session messages.",
          parameters: {
            type: "object",
            properties: { out: { type: "string" }, format: { type: "string" } }
          }
        }
      }
    ];
  }

  listToolNames(): string[] {
    return this.definitions().map((x) => x.function.name);
  }

  execute(name: string, rawArgs: unknown, sessionMessagesJson = ""): ToolResult {
    const args = parseArgs(rawArgs);
    if (name === "read_file") {
      const p = String(args.path ?? "");
      if (!p || !withinRoot(this.projectRoot, p) || !allowedPath(this.config, this.projectRoot, p)) {
        return { ok: false, output: "path denied" };
      }
      const full = path.resolve(this.projectRoot, p);
      if (!fs.existsSync(full)) {
        return { ok: false, output: "not found" };
      }
      return { ok: true, output: fs.readFileSync(full, "utf8") };
    }

    if (name === "write_file") {
      if (this.config.safe_mode === "strict") {
        return { ok: false, output: "write requires explicit approval", requiresApproval: true };
      }
      const p = String(args.path ?? "");
      const content = String(args.content ?? "");
      if (!p || !withinRoot(this.projectRoot, p) || !allowedPath(this.config, this.projectRoot, p)) {
        return { ok: false, output: "path denied" };
      }
      const full = path.resolve(this.projectRoot, p);
      fs.mkdirSync(path.dirname(full), { recursive: true });
      fs.writeFileSync(full, content, "utf8");
      return { ok: true, output: `wrote ${p}` };
    }

    if (name === "list_dir") {
      const p = String(args.path ?? ".");
      if (!withinRoot(this.projectRoot, p) || !allowedPath(this.config, this.projectRoot, p)) {
        return { ok: false, output: "path denied" };
      }
      const full = path.resolve(this.projectRoot, p);
      if (!fs.existsSync(full)) {
        return { ok: false, output: "not found" };
      }
      const entries = fs.readdirSync(full, { withFileTypes: true }).map((e) => (e.isDirectory() ? `${e.name}/` : e.name));
      return { ok: true, output: entries.join("\n") };
    }

    if (name === "grep_text") {
      const patternRaw = String(args.pattern ?? "");
      if (!patternRaw) {
        return { ok: false, output: "missing pattern" };
      }
      const start = String(args.path ?? ".");
      if (!withinRoot(this.projectRoot, start) || !allowedPath(this.config, this.projectRoot, start)) {
        return { ok: false, output: "path denied" };
      }
      const regex = new RegExp(patternRaw, "i");
      const root = path.resolve(this.projectRoot, start);
      const files: string[] = [];
      const walk = (dir: string): void => {
        for (const item of fs.readdirSync(dir, { withFileTypes: true })) {
          if (item.name === ".git" || item.name === "node_modules" || item.name === ".minicli4") {
            continue;
          }
          const full = path.join(dir, item.name);
          if (item.isDirectory()) {
            walk(full);
          } else {
            files.push(full);
          }
        }
      };
      walk(root);
      const matches: string[] = [];
      for (const f of files.slice(0, 500)) {
        let text = "";
        try {
          text = fs.readFileSync(f, "utf8");
        } catch {
          continue;
        }
        const lines = text.split(/\r?\n/);
        for (let i = 0; i < lines.length; i += 1) {
          if (regex.test(lines[i])) {
            matches.push(`${path.relative(this.projectRoot, f)}:${i + 1}: ${lines[i]}`);
            if (matches.length >= 50) {
              return { ok: true, output: matches.join("\n") };
            }
          }
        }
      }
      return { ok: true, output: matches.join("\n") || "no matches" };
    }

    if (name === "run_shell") {
      const command = String(args.command ?? "").trim();
      if (!command) {
        return { ok: false, output: "missing command" };
      }
      const head = command.split(/\s+/)[0];
      if (this.config.safe_mode === "strict" && !this.config.shell_allowlist.includes(head)) {
        return { ok: false, output: `command blocked in strict mode: ${head}`, requiresApproval: true };
      }
      const proc = childProcess.spawnSync(command, {
        shell: true,
        cwd: this.projectRoot,
        encoding: "utf8",
        timeout: Math.min(this.config.timeout_ms, 30_000)
      });
      const output = [proc.stdout ?? "", proc.stderr ?? ""].join("\n").trim();
      return { ok: proc.status === 0, output: output || `(exit ${proc.status ?? -1})` };
    }

    if (name === "project_tree") {
      const p = String(args.path ?? ".");
      const depth = Number(args.depth ?? 3);
      if (!withinRoot(this.projectRoot, p) || !allowedPath(this.config, this.projectRoot, p)) {
        return { ok: false, output: "path denied" };
      }
      const root = path.resolve(this.projectRoot, p);
      const lines: string[] = [];
      const visit = (dir: string, prefix: string, d: number): void => {
        if (d < 0) {
          return;
        }
        const entries = fs.readdirSync(dir, { withFileTypes: true }).slice(0, 50);
        for (const e of entries) {
          const line = `${prefix}${e.isDirectory() ? "[D]" : "[F]"} ${e.name}`;
          lines.push(line);
          if (e.isDirectory()) {
            visit(path.join(dir, e.name), `${prefix}  `, d - 1);
          }
        }
      };
      visit(root, "", Number.isFinite(depth) ? depth : 3);
      return { ok: true, output: lines.join("\n") };
    }

    if (name === "session_export") {
      const out = String(args.out ?? ".minicli4/sessions/export.json");
      if (!withinRoot(this.projectRoot, out)) {
        return { ok: false, output: "path denied" };
      }
      const full = path.resolve(this.projectRoot, out);
      fs.mkdirSync(path.dirname(full), { recursive: true });
      fs.writeFileSync(full, sessionMessagesJson, "utf8");
      return { ok: true, output: `exported ${out}` };
    }

    return { ok: false, output: `unknown tool: ${name}` };
  }
}

File: aicode/minicli4/cpp/src/tools.cpp
#include "minicli4/tools.hpp"

#include <algorithm>
#include <functional>
#include <fstream>
#include <regex>
#include <sstream>

namespace minicli4 {

ToolRegistry::ToolRegistry(std::filesystem::path project_root, RuntimeConfig config)
    : project_root_(std::move(project_root)), config_(std::move(config)) {}

std::vector<std::string> ToolRegistry::list_tool_names() const {
  return {"read_file", "write_file", "list_dir", "grep_text", "run_shell", "project_tree", "session_export"};
}

bool ToolRegistry::within_root(const std::string& target) const {
  const auto root = std::filesystem::weakly_canonical(project_root_);
  const auto resolved = std::filesystem::weakly_canonical(project_root_ / target);
  return resolved == root || resolved.string().rfind(root.string(), 0) == 0;
}

bool ToolRegistry::allowed_path(const std::string& target) const {
  const auto resolved = std::filesystem::weakly_canonical(project_root_ / target);
  for (const auto& p : config_.allowed_paths) {
    const auto allowed = std::filesystem::weakly_canonical(project_root_ / p);
    if (resolved == allowed || resolved.string().rfind(allowed.string(), 0) == 0) {
      return true;
    }
  }
  return false;
}

ToolResult ToolRegistry::execute(
    const std::string& name,
    const std::map<std::string, std::string>& args,
    const std::string& session_messages_json) const {
  if (name == "read_file") {
    const auto it = args.find("path");
    const std::string path = it == args.end() ? "" : it->second;
    if (path.empty() || !within_root(path) || !allowed_path(path)) {
      return {false, "path denied", false};
    }
    const auto full = std::filesystem::weakly_canonical(project_root_ / path);
    if (!std::filesystem::exists(full)) {
      return {false, "not found", false};
    }
    std::ifstream in(full, std::ios::binary);
    std::ostringstream ss;
    ss << in.rdbuf();
    return {true, ss.str(), false};
  }

  if (name == "write_file") {
    if (config_.safe_mode == "strict") {
      return {false, "write requires explicit approval", true};
    }
    const std::string path = args.count("path") ? args.at("path") : "";
    const std::string content = args.count("content") ? args.at("content") : "";
    if (path.empty() || !within_root(path) || !allowed_path(path)) {
      return {false, "path denied", false};
    }
    const auto full = std::filesystem::weakly_canonical(project_root_ / path);
    std::filesystem::create_directories(full.parent_path());
    std::ofstream out(full, std::ios::binary);
    out << content;
    return {true, "wrote " + path, false};
  }

  if (name == "list_dir") {
    const std::string path = args.count("path") ? args.at("path") : ".";
    if (!within_root(path) || !allowed_path(path)) {
      return {false, "path denied", false};
    }
    const auto full = std::filesystem::weakly_canonical(project_root_ / path);
    if (!std::filesystem::exists(full)) {
      return {false, "not found", false};
    }
    std::vector<std::string> entries;
    for (const auto& entry : std::filesystem::directory_iterator(full)) {
      entries.push_back(entry.path().filename().string() + (entry.is_directory() ? "/" : ""));
    }
    std::sort(entries.begin(), entries.end());
    std::ostringstream out;
    for (size_t i = 0; i < entries.size(); i++) {
      out << entries[i];
      if (i + 1 < entries.size()) {
        out << "\n";
      }
    }
    return {true, out.str(), false};
  }

  if (name == "grep_text") {
    const std::string pattern = args.count("pattern") ? args.at("pattern") : "";
    const std::string start = args.count("path") ? args.at("path") : ".";
    if (pattern.empty()) {
      return {false, "missing pattern", false};
    }
    if (!within_root(start) || !allowed_path(start)) {
      return {false, "path denied", false};
    }
    std::regex regex;
    try {
      regex = std::regex(pattern, std::regex::icase);
    } catch (...) {
      return {false, "invalid regex", false};
    }
    std::vector<std::string> matches;
    const auto root = std::filesystem::weakly_canonical(project_root_ / start);
    for (const auto& entry : std::filesystem::recursive_directory_iterator(root)) {
      const auto rel = std::filesystem::relative(entry.path(), project_root_).generic_string();
      if (rel.find(".git/") == 0 || rel.find("node_modules/") == 0 || rel.find(".minicli4/") == 0) {
        continue;
      }
      if (!entry.is_regular_file()) {
        continue;
      }
      std::ifstream in(entry.path());
      if (!in.is_open()) {
        continue;
      }
      std::string line;
      int line_no = 0;
      while (std::getline(in, line)) {
        line_no += 1;
        if (std::regex_search(line, regex)) {
          std::ostringstream one;
          one << rel << ":" << line_no << ": " << line;
          matches.push_back(one.str());
          if (matches.size() >= 50) {
            break;
          }
        }
      }
      if (matches.size() >= 50) {
        break;
      }
    }
    if (matches.empty()) {
      return {true, "no matches", false};
    }
    std::ostringstream out;
    for (size_t i = 0; i < matches.size(); i++) {
      out << matches[i];
      if (i + 1 < matches.size()) {
        out << "\n";
      }
    }
    return {true, out.str(), false};
  }

  if (name == "run_shell") {
    const std::string cmd = args.count("command") ? args.at("command") : "";
    if (cmd.empty()) {
      return {false, "missing command", false};
    }
    const auto space = cmd.find(' ');
    const std::string head = cmd.substr(0, space == std::string::npos ? cmd.size() : space);
    if (config_.safe_mode == "strict") {
      if (std::find(config_.shell_allowlist.begin(), config_.shell_allowlist.end(), head) == config_.shell_allowlist.end()) {
        return {false, "command blocked in strict mode: " + head, true};
      }
    }
    return {true, "(execution omitted in tests)", false};
  }

  if (name == "project_tree") {
    const std::string path = args.count("path") ? args.at("path") : ".";
    int depth = 3;
    if (args.count("depth")) {
      depth = std::max(0, std::stoi(args.at("depth")));
    }
    if (!within_root(path) || !allowed_path(path)) {
      return {false, "path denied", false};
    }
    const auto root = std::filesystem::weakly_canonical(project_root_ / path);
    std::ostringstream out;
    std::function<void(const std::filesystem::path&, int, const std::string&)> walk =
        [&](const std::filesystem::path& dir, int d, const std::string& prefix) {
          if (d < 0) {
            return;
          }
          std::vector<std::filesystem::path> items;
          for (const auto& entry : std::filesystem::directory_iterator(dir)) {
            items.push_back(entry.path());
          }
          std::sort(items.begin(), items.end());
          for (const auto& p : items) {
            const bool is_dir = std::filesystem::is_directory(p);
            out << prefix << (is_dir ? "[D] " : "[F] ") << p.filename().string() << "\n";
            if (is_dir) {
              walk(p, d - 1, prefix + "  ");
            }
          }
        };
    walk(root, depth, "");
    return {true, out.str(), false};
  }

  if (name == "session_export") {
    const std::string out_path = args.count("out") ? args.at("out") : ".minicli4/sessions/export.json";
    if (!within_root(out_path)) {
      return {false, "path denied", false};
    }
    const auto full = std::filesystem::weakly_canonical(project_root_ / out_path);
    std::filesystem::create_directories(full.parent_path());
    std::ofstream out(full, std::ios::binary);
    out << session_messages_json;
    return {true, "exported " + out_path, false};
  }

  return {false, "unknown tool: " + name, false};
}

}  // namespace minicli4
